x/gamm/keeper/msg_server_test.go

```diff
@@ -356,3 +362,211 @@ func (suite *KeeperTestSuite) TestExitPool_Events() {
+		}
+
+		// Note cl pool balance before migration
+		clPoolEthBalanceBeforeMigration := suite.App.BankKeeper.GetBalance(suite.Ctx, clPoolAddress, ETH)
+		clPoolUsdcBalanceBeforeMigration := suite.App.BankKeeper.GetBalance(suite.Ctx, clPoolAddress, USDC)
+
+		// Reset event counts to 0 by creating a new manager.
+		suite.Ctx = suite.Ctx.WithEventManager(sdk.NewEventManager())
+		suite.Require().Equal(0, len(suite.Ctx.EventManager().Events()))
+
+		// Migrate the user's gamm shares to a full range concentrated liquidity position
+		resp, err := msgServer.MigrateSharesToFullRangeConcentratedPosition(sdk.WrapSDKContext(suite.Ctx), msg)
+		if test.expectedErr != nil {
+			suite.Require().Error(err)
+			suite.Require().ErrorContains(err, test.expectedErr.Error())
+			continue
+		}
+
+		suite.Require().NoError(err)
+
+		// Assert events are emitted
+		suite.AssertEventEmitted(suite.Ctx, types.TypeEvtMigrateShares, test.expectedMigrateShareEvents)
+		suite.AssertEventEmitted(suite.Ctx, sdk.EventTypeMessage, test.expectedMessageEvents)
+
+		// Note gamm pool balance after migration
+		gammPoolEthBalancePostMigrate := suite.App.BankKeeper.GetBalance(suite.Ctx, gammPoolAddress, ETH)
+		gammPoolUsdcBalancePostMigrate := suite.App.BankKeeper.GetBalance(suite.Ctx, gammPoolAddress, USDC)
+
+		// Note user balance after migration
+		userEthBalanceTransferredToClPool := gammPoolEthBalancePostJoin.Sub(gammPoolEthBalancePostMigrate)
+		userUsdcBalanceTransferredToClPool := gammPoolUsdcBalancePostJoin.Sub(gammPoolUsdcBalancePostMigrate)
+
+		// Note cl pool balance after migration
+		clPoolEthBalanceAfterMigration := suite.App.BankKeeper.GetBalance(suite.Ctx, clPoolAddress, ETH)
+		clPoolUsdcBalanceAfterMigration := suite.App.BankKeeper.GetBalance(suite.Ctx, clPoolAddress, USDC)
+
+		// The balance in the cl pool should be equal to what the user previously had in the gamm pool
+		poolEthBalanceInTolerance := userEthBalanceTransferredToClPool.Amount.Sub(clPoolEthBalanceAfterMigration.Amount).LTE(sdk.NewInt(1))
+		poolUsdcBalanceInTolerance := userEthBalanceTransferredToClPool.Amount.Sub(clPoolEthBalanceAfterMigration.Amount).LTE(sdk.NewInt(1))
+		fmt.Printf("userEthBalanceTransferredToClPool.Amount %v \n", userEthBalanceTransferredToClPool.Amount)
+		fmt.Printf("clPoolEthBalanceAfterMigration.Amount %v \n", clPoolEthBalanceAfterMigration.Amount)
+		fmt.Printf("userUsdcBalanceTransferredToClPool.Amount %v \n", userUsdcBalanceTransferredToClPool.Amount)
+		fmt.Printf("clPoolUsdcBalanceAfterMigration.Amount %v \n", clPoolUsdcBalanceAfterMigration.Amount)
+		fmt.Printf("poolUsdcBalanceInTolerance %v \n", poolUsdcBalanceInTolerance)
+		suite.Require().True(poolEthBalanceInTolerance)
+		suite.Require().True(poolUsdcBalanceInTolerance)
+		// suite.Require().Equal(userEthBalanceTransferredToClPool.String(), clPoolEthBalanceAfterMigration.String())
+		// suite.Require().Equal(userUsdcBalanceTransferredToClPool.String(), clPoolUsdcBalanceAfterMigration.String())
+
```

Comment:
Can we add checks for user's postion and balance before withdrawing?

Reply:
https://github.com/osmosis-labs/osmosis/pull/4051/commits/de1590c38fb8c5f801f5528677464682a79840e3